File: Haskell2/Problem5.hs
perfect :: Int -> Bool
perfect n = n == sum [x | x <- [1..n], n `mod` x == 0, x /= n]
mullist xs = [x | x <- [1..], isProd xs x]

isProd [] _ = False
isProd (x:xs) n =  if n == 1 && x /= 1 then False
					else if x == 1 then True
					else if n `mod` x == 0 then isProdRecurse (x:xs) (quot n x)
					else isProdRecurse xs n
isProdRecurse _ 1 = True
isProdRecurse [] _ = False
isProdRecurse (x:xs) n = if n `mod` x == 0 then isProdRecurse (x:xs) (quot n x)
					else isProdRecurse xs n

File: Prolog1/Problem5.pl
hailstone(X, [X], 1).
hailstone(X,[X|S], N) :- 0 is X mod 2, X1 is X / 2, N > 0, X > 0, N1 is N - 1, hailstone(X1,S, N1).
hailstone(X,[X|S], N) :- 1 is X mod 2, X1 is (3 * X) + 1, N > 0, X > 0, N1 is N - 1, hailstone(X1, S, N1).